frontend/catalyst/compiler.py

```diff
@@ -478,19 +478,81 @@ def run_from_ir(self, ir: str, module_name: str, workspace: Directory):
 
         return output_object_name, out_IR
 
+    def has_xdsl_passes_in_transform_modules(self, mlir_module):
+        """Check if the MLIR module contains xDSL passes in transform dialect.
+
+        This checks for the 'uses_xdsl_passes' attribute that is set during
+        lowering on transform modules when xDSL passes are added to the transform pipeline.
+
+        Args:
+            mlir_module: MLIR module to check for xDSL passes
+
+        Returns:
+            bool: True if xDSL passes detected in any transform module
+        """
+
+        def has_both_attributes(attrs):
+            """Check if attributes dict has both required keys."""
+            try:
+                has_transform = (
+                    "transform.with_named_sequence" in attrs
+                    or "transform.with_named_sequence" in getattr(attrs, "keys", lambda: [])()
+                )
+                has_xdsl = (
+                    "uses_xdsl_passes" in attrs
+                    or "uses_xdsl_passes" in getattr(attrs, "keys", lambda: [])()
+                )
+                return has_transform and has_xdsl
+            except (AttributeError, KeyError, TypeError):
+                return False
+
+        try:
+            for op in mlir_module.operation.regions[0].blocks[0].operations:
+                try:
+                    # Check nested modules (look for transform modules)
+                    if hasattr(op, "regions") and len(op.regions) > 0:
+                        for nested_op in op.regions[0].blocks[0].operations:
+                            if hasattr(nested_op, "attributes"):
+                                if has_both_attributes(nested_op.attributes):
+                                    return True
```

Comment:
I think this should also work (and it avoids an extra level of indentation):

```suggestion
                            if hasattr(nested_op, "attributes") and has_both_attributes(nested_op.attributes):
                                return True
```